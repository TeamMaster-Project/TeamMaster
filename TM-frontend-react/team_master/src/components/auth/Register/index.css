.register-form-container {
  background-color: #1029b8;

  min-height: 94.3vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: 1.05rem;
}

.register-form-card {
  background-color: #fff;
  border-radius: 15px;
  width: 50vw;
  opacity: 0.9;
}
.register-form-card:hover {
  opacity: 1;
}

.register-form {
  padding: 25px;
}
.register-form:hover {
  padding: 30px;
}

.form-submit-button {
  text-align: center;
}


.login,
.image {
  min-height: 94vh;
}

.bg-image-register {
  background-image: url('../../../assets/images/hands-typing.jpg');
  background-size: cover;
  background-position: center center;
}