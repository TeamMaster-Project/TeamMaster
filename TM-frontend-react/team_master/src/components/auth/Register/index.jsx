import React from "react";
import Form from "../../Common/form";
import Joi from "joi-browser";
import * as userService from "../../../services/userService";
import auth from "../../../services/authService";
import "./index.css";
import { addUsers } from "../../../services/chatboxService";
import { toast } from "react-toastify";
import { Link } from "react-router-dom";
import Loader from "../../PreLoader/Loader";
import Logo from "../../../assets/images/Logo/TMLogo1.png"


class Register extends Form {
  state = {
    data: { username: "", password: "", name: "" },
    errors: {},
    isLoading: false,
    isSuccess: false,
    isFail: false,
  };

  schema = {
    //Joi validate schema
    username: Joi.string().required().label("Username"),
    password: Joi.string().required().min(5).label("Password"),
    name: Joi.string().required().min(5).label("Name"),
  };

  createUserInChatEngine = async (currentUser) => {
    try{
      await addUsers(currentUser);
    }catch(error){
      console.log(error);
      toast("User Registration Failed in ChatEngine")
    }
      
  };

  async componentDidMount() {
      if(this.props.data)
        this.setState({data: this.props.data})

      this.setState({isLoading: true});
        setTimeout(()=>{
            this.setState({isLoading: false});
        },2000)
  }

  doSubmit = async () => {
    try {
      this.setState({isLoading: true})
      const response = await userService.register(this.state.data);//Register user
      await this.createUserInChatEngine(response.data);//Register user in ChatEngine

      //Log in user soon after registration
      //Redirecting to main home page to get current user logged in
      auth.loginWithJwt("token", response.headers["x-auth-token"]); //Store the token in localStorage when after creating a new user
      //window.location = "/"; //full Reload and redirecting to homepage
      this.setState({isLoading: false, isSuccess: true})
      this.props.history.push("/newproject"); 
      toast("User Registered Successfully");
    } catch (ex) {
      this.setState({isLoading: false, isFail: true})
      if (ex.response && ex.response.status === 400) {
        //Handle user already registered error
        const errors = { ...this.state.errors };
        errors.username = ex.response.data;
        this.setState({ errors });
        toast.error("Registration fail");
      }
    }
  };
  render() {

    return (
      <div class="container-fluid">
          <div class="row no-gutter">
            {/* left Half */}
            <div class="col-md-6 d-none d-md-flex bg-image-register"></div>

            {/* Right Half */}
            <div class="col-md-6 bg-light">
              <div class="login d-flex align-items-center py-5">
                  <div class="container">
                    <div class="row">
                        <div class="col-lg-10 col-xl-7 mx-auto">
                            <h3><img src={Logo} width="110px" height="100px"/></h3>
                            <h3 class="display-4">Get an account</h3>
                            <h6 class="text-muted mb-4">Enter your email and password</h6>
                            {this.state.isLoading ? ( 
                              <h6>Please wait...
                                  <Loader/>
                              </h6>
                              ) : null 
                            }
                            { !this.state.isLoading && this.state.isSuccess ? <h6 style={{color: 'green'}}>Registration success</h6>: null }
                            { !this.state.isLoading && this.state.isFail ? <h6 style={{color: 'red'}}>Registration fail</h6>: null }
                             <div className="register-form">
                               <form onSubmit={this.handleSubmit}>
                                  {this.renderInputs("username", "Email / Username", "text")}
                                  {this.renderInputs("password", "Password", "password")}
                                  {/* name label datatype */}
                                  {this.renderInputs("name", "Name", "text")}
                                  <div className="form-submit-button">
                                    {this.renderButton("Register")}
                                  </div>
                                </form>
                              </div>
                               <h6 class="text-muted mb-4"> Already have an account ? &nbsp;
                                  <Link
                                  to="/login"
                                  className="mb-4"
                                >
                                 <i>Log in</i>
                                </Link>
                               </h6>
                        </div>
                    </div>
                  </div>
              </div>
            </div>
          </div>
        </div>
    );
  }
}

export default Register;
