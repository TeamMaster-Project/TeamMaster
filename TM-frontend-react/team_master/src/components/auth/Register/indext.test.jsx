import React from 'react'
import {render, screen, cleanup, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Register from '.';

afterEach(()=>{
    cleanup();
})

const windowOpen = jest.fn();
window.open = windowOpen;

const RegisterFormPropsValid = {
    data: { username: "newuser@gmail.com", password: "123456789", name: "Team Master Test User" },
    doSubmit: jest.fn(),
};

const RegisterFormPropsWithExistingUsername = {
    data: { username: "user@gmail.com", password: "123456789", name: "Team Master Test User" },
    doSubmit: jest.fn(),
};


test('Matches Snapshot',()=>{
    const tree = renderer.create(<Register/>).toJSON();
    expect(tree).toMatchSnapshot();
})

describe('Testing UI', () => {
    it('Form elements rendering correctly', async () => {
        const { getByTestId } = render(<Register/>);

        const emailInput = getByTestId("username");
        expect(emailInput).toBeInTheDocument();

        const passwordInput = getByTestId("password");
        expect(passwordInput).toBeInTheDocument()

        const nameInput = getByTestId("name");
        expect(nameInput).toBeInTheDocument()

        expect(screen.getByRole('button', { name: 'Register' })).toBeInTheDocument();

    });  
});

describe('Testing registration with different combinations of inputs', () => {

    // This test case skipped because of trying to create a new user every time the test case run
    it.skip('Sign in with correct / new credentials', async () => {
        const { getByRole } = render(<Register {...RegisterFormPropsValid} />);
        const RegisterButton = getByRole('button', { name: 'Register' });
        fireEvent.click(RegisterButton);
        await screen.findByText("Please wait...");
        screen.findByText("Registration success");
    });  

    it('Trying to Sign in with correct an existing registered user', async () => {
        const { getByRole } = render(<Register {...RegisterFormPropsWithExistingUsername} />);
        const RegisterButton = getByRole('button', { name: 'Register' });
        fireEvent.click(RegisterButton);
        await screen.findByText("Please wait...");
        screen.findByText("Registration fail");
    });  
})