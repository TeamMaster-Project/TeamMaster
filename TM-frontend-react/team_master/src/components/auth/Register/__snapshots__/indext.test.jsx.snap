// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
<div
  class="container-fluid"
>
  <div
    class="row no-gutter"
  >
    <div
      class="col-md-6 d-none d-md-flex bg-image-register"
    />
    <div
      class="col-md-6 bg-light"
    >
      <div
        class="login d-flex align-items-center py-5"
      >
        <div
          class="container"
        >
          <div
            class="row"
          >
            <div
              class="col-lg-10 col-xl-7 mx-auto"
            >
              <h3>
                <img
                  height="100px"
                  src="TMLogo1.png"
                  width="110px"
                />
              </h3>
              <h3
                class="display-4"
              >
                Get an account
              </h3>
              <h6
                class="text-muted mb-4"
              >
                Enter your email and password
              </h6>
              <h6>
                Please wait...
                <div
                  className="container"
                >
                  <h2>
                    <img
                      height="100px"
                      src="loader1.gif"
                      width="100px"
                    />
                  </h2>
                </div>
              </h6>
              <div
                className="register-form"
              >
                <form
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="username"
                    >
                      Email / Username
                    </label>
                    <input
                      className="form-control"
                      data-testid="username"
                      id="username"
                      name="username"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="password"
                    >
                      Password
                    </label>
                    <input
                      className="form-control"
                      data-testid="password"
                      id="password"
                      name="password"
                      onChange={[Function]}
                      type="password"
                      value=""
                    />
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="name"
                    >
                      Name
                    </label>
                    <input
                      className="form-control"
                      data-testid="name"
                      id="name"
                      name="name"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    className="form-submit-button"
                  >
                    <button
                      className="btn btn-primary"
                      data-testid="Register"
                      disabled={
                        Object {
                          "name": "\\"Name\\" length must be at least 5 characters long",
                          "password": "\\"Password\\" length must be at least 5 characters long",
                          "username": "\\"Username\\" is not allowed to be empty",
                        }
                      }
                    >
                      Register
                    </button>
                  </div>
                </form>
              </div>
              <h6
                class="text-muted mb-4"
              >
                 Already have an account ? Â 
                <a
                  className="mb-4"
                  href="/login"
                  onClick={[Function]}
                >
                  <i>
                    Log in
                  </i>
                </a>
              </h6>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
